### PR TITLE
Implemented batching for Inception example.

### PR DESCRIPTION
**Pull request by eriophora**
Implemented batching for the inception example with TensorFlow Serving. This was accomplished by performing client-side preprocessing, using PIL. It appears, however, that this requires a newer version of PIL than is provided with TensorFlow, and it has to be imported before TensorFlow to avoid bizarre filestream issues with PIL. 

Additionally, three types of preprocessing were implemented:
- Resize (the default -- and original way)
- Crop (the image is resized to preserve the maximum amount of image possible then cropped to the correct size)
- Pad & Resize (the image is symmetrically padded to the correct aspect ratio then resized)

See issue #38

### CODE REVIEW

**Comment by eriophora** [comment]
I signed it!

**Comment by vinuraja** [comment]
@fangweili could you please take a look at this ?

File: tensorflow_serving/example/inception_inference.cc [hunk 1]
```diff
@@ -208,10 +212,9 @@ InceptionServiceImpl::InceptionServiceImpl(
   // the numbers are extremely performance critical and should be tuned based
   // specific graph structure and usage.
   tensorflow::serving::StreamingBatchScheduler<Task>::Options scheduler_options;
-  scheduler_options.thread_pool_name = "inception_service_batch_threads";
-  // TODO(27776734): Current exported model supports only batch_size=1
-  // See inception_export.py for details.
-  scheduler_options.max_batch_size = 1;
+  scheduler_options.batch_timeout_micros = 1000 * 1000;  // 1 second
+  scheduler_options.num_batch_threads = 4;
+  scheduler_options.max_batch_size = 22;
```
Comment (chrisolston) [inline tensorflow_serving/example/inception_inference.cc]: Maybe add a comment on how you selected 22 as the max batch size?

Similar for num_batch_threads: are you aiming to match the server machine's number of GPU devices or ??

**Comment by eriophora** [comment]
Actually, it was simply the result of testing. This was tested on an Amazon EC2 GPU g2.2xlarge. The original number of images, 32, is too large for these machines and they will repeatedly emit those errors--I can't remember exactly what they are, orphaned or ejected memory or somesuch--and then crash.

The number of threads is somewhat arbitrary, but having at least 4 helps to avoid the "no available batches" as it appears that TF Serving will start evaluating a batch that's not full but that has timed out even if the compute resource is not yet free. 

However, to get the whole thing working on EC2 required quite a bit of modification -- and as such this could certainly benefit from some testing on a 'normal' set up.

**Comment by sukritiramesh** [comment]
@fangweili Can you please take a look at the PR? Thanks!

**Comment by fangweili** [comment]
@eriophora Sorry for the late response. This looks really good. I wonder if you can fork files into inception_batching_* instead and submit as a separate example. We would like to keep the original inception example as simple as possible, for tutorial purpose. You can update the TODO in inception_export.py to point to this new example. Thanks a lot!

File: tensorflow_serving/example/inception_inference.proto [hunk 2]
```diff
@@ -1,12 +1,11 @@
...
-
 syntax = "proto3";
 
 package tensorflow.serving;
 
 message InceptionRequest {
-  // JPEG encoded stream of the image to be classified.
-  bytes jpeg_encoded = 1;
+  // Image data as a flattened array of floats
+  repeated float image_data = 1 [packed = true];
```
Comment (rcompton) [inline tensorflow_serving/example/inception_inference.proto]: is this necessary for batching?

**Comment by chrisolston** [comment]
This PR has been sitting for a while now. I'm going to close it due to inactivity. Feel free to reopen.

**Pull request closed without merging**